core/src/main/java/com/sequenceiq/cloudbreak/service/stack/flow/StackOperationService.java

```diff
@@ -113,54 +113,57 @@ public FlowIdentifier removeInstance(Stack stack, String instanceId, boolean for
-                LOGGER.info("ZZZ: removeInstances: computedInstanceIds: {}", instanceIdsByHostgroupMap);
-                // TODO CB-14929: Validations for start-stop
-                if (!forced) {
-                    for (Entry<String, Set<Long>> entry : instanceIdsByHostgroupMap.entrySet()) {
-                        String instanceGroupName = entry.getKey();
-                        int scalingAdjustment = entry.getValue().size() * -1;
-                        updateNodeCountValidator.validateScalabilityOfInstanceGroup(stack, instanceGroupName, scalingAdjustment);
-                        updateNodeCountValidator.validateScalingAdjustment(instanceGroupName, scalingAdjustment, stack);
-                    }
-                }
-                if (scalingStrategy != null) {
-                    // TODO: CB-14929: CB-15154 stopstart can only support a single hostGroup being specified.
-                    if (instanceIdsByHostgroupMap.size() > 1) {
-                        throw new BadRequestException("Downscale via Instance Stop cannot process more than one host group");
-                    }
-                    stackUpdater.updateStackStatus(stack.getId(), DetailedStackStatus.DOWNSCALE_BY_STOP_REQUESTED,
-                            "Requested node count for downscaling (stopstart): " + instanceIds.size());
-                    return flowManager.triggerStopStartStackDownscale(stack.getId(), instanceIdsByHostgroupMap, forced);
-                } else {
-                    return flowManager.triggerStackRemoveInstances(stack.getId(), instanceIdsByHostgroupMap, forced);
-                }
-            });
-        } catch (TransactionExecutionException e) {
-            if (e.getCause() instanceof BadRequestException) {
-                throw e.getCause();
+    public FlowIdentifier stopInstances(Stack stack, Collection<String> instanceIds, boolean forced) {
+        LOGGER.info("Received stop instances request for instanceIds: [{}]", instanceIds);
+
+        if (instanceIds == null || instanceIds.isEmpty()) {
+            throw new BadRequestException("Stop request cannot process an empty instanceIds collection");
+        }
+        Map<String, Set<Long>> instanceIdsByHostgroupMap = new HashMap<>();
+        for (String instanceId : instanceIds) {
+            InstanceMetaData metaData = updateNodeCountValidator.validateInstanceForStop(instanceId, stack);
+            if (metaData != null) {
+                instanceIdsByHostgroupMap.computeIfAbsent(metaData.getInstanceGroupName(), s -> new LinkedHashSet<>()).add(metaData.getPrivateId());
             }
-            throw new TransactionRuntimeExecutionException(e);
         }
+        if (instanceIdsByHostgroupMap.size() > 1) {
+            throw new BadRequestException("Downscale via Instance Stop cannot process more than one host group");
+        }
+        updateNodeCountValidator.validateServiceRoles(stack, instanceIdsByHostgroupMap.entrySet()
```

Comment:
This also validates instance counts. In a future jira, it would be useful to differentiate between roles which should not go to 0 (including a STOP) - e.g. DATANODES. vs roles which can go to 0 with STOP, but not with an actual downscale (e.g. NODEMANAGERS). The 2nd category is a CM side restriction more than an actual service limitation / data loss scenario.

Reply:
Agreed. We test with default templates so the NODEMANAGER never goes to 0.